core/platform/auth/firebase_auth_services.py

```diff
@@ -328,8 +350,15 @@ def associate_multi_auth_ids_to_user_ids(auth_id_user_id_pairs):
 
     assoc_by_user_id_models = [
         auth_models.UserAuthDetailsModel(id=user_id, firebase_auth_id=auth_id)
-        for auth_id, user_id in python_utils.ZIP(auth_ids, user_ids)
+        for auth_id, user_id, assoc_by_user_id_model in python_utils.ZIP(
+            auth_ids, user_ids,
+            auth_models.UserAuthDetailsModel.get_multi(user_ids))
+        if (assoc_by_user_id_model is None or
```

Comment:
I don't think I understand this "if" clause. If you've validated above that the IDs are not already associated, then why do we need this guard here?

Reply:
This was a compromise I had to make because of  `user_services`'s dependency on `UserAuthDetailsModel` to manage full/profile users (specifically the `parent_user_id` property.
-   I _do_ want to keep `UserAuthDetailsModel` management in the platform layer, because the auth_id property we read from/write to depends on the platform (GAE → `gae_id`, Firebase → `firebase_auth_id`).
-   I _do not_ want to lift `parent_user_id` management to the platform layer, because it doesn't belong here.

Incidentally, this is what sparked the conversations between @BenHenning, you, me.

Reply:
Ack'd, let's discuss in the other comment thread: https://github.com/oppia/oppia/pull/11588#discussion_r554284761.

Reply:
Hm, could you maybe add some comments in the code to explain what's going on (here and perhaps for other places where I've left comments)? Not so much in terms of the platform layer vs stuff at the top, but mainly to explain why certain things are needed (i.e., what bad things will happen if those clauses are removed).

I'm trying to make sure devs don't remove them by accident, and that they understand why they should not remove them. From looking at the code I still don't quite understand what this has to do with parent_user_id exactly.

Reply:
OK, I've given it a shot. WDYT?

Reply:
I've made a bunch of updates here, PTAL

Reply:
I saw this thread earlier, but I didn't mention then: I don't actually have anything to add beyond your comments @brianrodri. (I didn't actually interpret @seanlip's comment as suggesting changes before, just clarification so I didn't think anything additional was needed from me). Please let me know if I should provide any additional context here.

The relationship being established in UserAuthModel otherwise looks good to me. I have no additional comments to add there.

Reply:
Have confirmed offline w/ @seanlip that this can be resolved, have added more comments and an extra check to make it more obvious why this is happening. Thanks Ben!